sync/src/main/kotlin/gropius/sync/AbstractSync.kt

```diff
@@ -510,16 +596,53 @@ abstract class AbstractSync(
     private suspend fun syncOutgoingLabels(
         timeline: List<TimelineItem>, imsProject: IMSProject, issueInfo: IssueConversionInformation
     ) {
-        val groups = timeline.filter { (it is AddedLabelEvent) || (it is RemovedLabelEvent) }.groupBy {
-            when (it) {
-                is AddedLabelEvent -> it.addedLabel().value
-                is RemovedLabelEvent -> it.removedLabel().value
-                else -> throw IllegalStateException()
+        val labelStateMap = this.labelStateMap(imsProject)
+        val stateLabelMap = labelStateMap.map { it.value to it.key }.toMap()
+        val virtualIDs = mutableMapOf<TimelineItem, String>()
+
+        val groups =
+            (timeline.filter { (it is AddedLabelEvent) || (it is RemovedLabelEvent) } + timeline.filterIsInstance<StateChangedEvent>()
+                .flatMap {
+                    val ret = mutableListOf<TimelineItem>()
+                    val labels = labelStateMap.mapNotNull {
+                        val labelName = it.key
+                        imsProject.trackable().value.labels().firstOrNull { it.name == labelName }
+                    }.toMutableSet()
+                    if (stateLabelMap.containsKey(it.newState().value.rawId!!)) {
+                        val name = stateLabelMap[it.newState().value.rawId!!]
+                        val label = imsProject.trackable().value.labels().firstOrNull { it.name == name }
+                        if (label != null) {
+                            labels -= label
+                            val elem = AddedLabelEvent(it.createdAt, it.lastModifiedAt)
+                            elem.createdBy().value = it.createdBy().value
+                            elem.issue().value = it.issue().value
+                            elem.lastModifiedBy().value = it.lastModifiedBy().value
+                            elem.addedLabel().value = label
+                            ret.add(elem)
+                            virtualIDs[elem] = it.rawId!! + "-added"
+                        }
+                    }
+                    for (label in labels) {
+                        val elem = RemovedLabelEvent(it.createdAt, it.lastModifiedAt)
+                        elem.createdBy().value = it.createdBy().value
+                        elem.issue().value = it.issue().value
+                        elem.lastModifiedBy().value = it.lastModifiedBy().value
+                        elem.removedLabel().value = label
+                        ret.add(elem)
+                        virtualIDs[elem] = it.rawId!! + "-" + label.rawId!! + "-removed"
+                    }
+                    ret
+                }).groupBy {
+                when (it) {
+                    is AddedLabelEvent -> it.addedLabel().value!!
+                    is RemovedLabelEvent -> it.removedLabel().value!!
+                    else -> throw IllegalStateException()
```

Comment:
maybe add an error message

Reply:
This error can only be triggered by a _significant_ programming error and in this case you have to search the source anyway

Reply:
99514a685576709fd9d23f887632e05931cf0aa1